### PR TITLE
Create 2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md

### PR DESCRIPTION
**Pull request by JoshMeredith**

### CODE REVIEW

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  get :: GetS a
+  putList :: [a] -> PutS
+  putList = putListOf put
+  getList :: GetS [a]
+  getList = getListOf get
+```
+
+Here we see that GHCJS has opted for a different approach that avoids the mutable buffer by instead using `Data.Binary` instances that work via concatenating lazy `ByteString`s. Additionally, the mutable tables are replaced with a `State` monad that holds the symbol table as a `Map` structure.
+
+Because `Data.Binary` forms lazy `ByteString`s, it's trivial to serialise the individual parts of the interface file and later concatenate these using `ByteString`'s monoid instance:
```
Comment (doyougnu) [inline blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md:127]: Again, you need to explain this code. I think the explanation you're missing is:

> Notice that with the monoid instance the code reads _in the order_ the bytestring is laid out: `hdr` then `symbs` then...Contrast this with the GHC approach; by looking at that code it is unclear what the exact order of the binary layout is.

or something like that

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+## Comparison
+
+### Advantages of the GHC Binary Approach
+
+* Interface file generation appears to be very fast: speculatively looking at the generated core, this could be because the optimiser has a much better time with the style of IO code that is used - rather than a limitation of more abstacted approaches
+* A significant amount of infrastructure already exists within GHC for interface files, so we can remove a lot of code used for `Objectable` and its maintenance burden along with it.
+
+### Advantages of the Objectable Approach
+
+* Avoids a lot of the weakly-typed binary buffer fiddling by using standard abstactions of `ByteString`s and `State` monads
```
Comment (doyougnu) [inline blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md:154]: Good! But relate this to a larger point: `we gain safety by utilizing the type system.`

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+* Concatenation of `ByteString`s allows a more declarative implementation of the serialised file's structure
+* Concatenation of `ByteString`s allows a more flexible ordering of the individual parts of the file
+* Don't need to carry around read/write functions (which are `undefined` during each-other's stage).
+
```
Comment (doyougnu) [inline blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md:158]: No disadvantages for either of the approaches? This comparison isn't complete without disadvantages right?

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+## Conclusion
+
+While making use of GHC's existing infrastructure lets the GHCJS backend to make use of the `FastString` and `Name` data types, as well as allowing for the removal of a significant amount of now-redundant code, the comparison provided the GHCJS's old approach makes it clear that GHC's `Binary` implementation, while very useful, could be improved in both readability and extensiblity.
```
[the remark this answers is not in the document]
Comment (JoshMeredith) [reply blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md:161]: I think I agree with you as this is written. I'll try and reword this to properly convey that I believe that this is less the result of an engineering trade-off, and more the result of legacy code.

I've previously found that hiding the handle in a reader did have this result (which you've identified as missing in this post) of causing the optimiser to have trouble simplifying the core, but I suspect that `Data.Binary` would have a lot of work done in its backend to support performance and so maybe this section would be better labelled as a sort of future work section - perhaps evaluating a proof of concept implementation of GHC's binary using this.

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md (hunk 3 above)
Comment (JoshMeredith) [reply blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md:158]: I had some trouble writing these in the sense that some of these points appear in both (e.g. objectable has the advantage of being strongly typed, and the GHC approach weakly enforces its types). Really, for the purpose of comparing two approaches, I don't know if I feel that it's worth mentioning an advantage of one approach if it's not a disadvantage of the other.

For clarity purposes, would you suggest mirroring these points (such as the strong/weak typing one)?

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
+# Objectable vs GHC Binary
+
+As part of the integration of GHCJS into GHC as a cross-compilation backend, we've converted the binary serialisation that GHCJS previously used, which was via its `Objectable` typeclass, into GHC's internal `Binary` typeclass representation.
+
+This allows us to make use of certain data types internal to GHC, i.e. `Name` and `FastString`, within GHCJS's ASTs, as well as maintaining consistency with GHC's `.hi` interface file serialisation.
```
[the remark this answers is not in the document]
Comment (JoshMeredith) [reply blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md]: Explained in more detail in the latest version

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+## How GHC Binary Works
+
+In GHC's `.hi` interface files, `Name` and `FastString` are serialised in a different way to other data structures, in that they a written in the main data structure payload as indicies of a table, and these tables contain the actual string-like data of these types. So, an interface file might resemble:
```
[the remark this answers is not in the document]
Comment (JoshMeredith) [reply blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md]: Based a paragraph in the latest version on this to clarify the use of these types.

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+   put_ bh name =
+      case getUserData bh of
+        UserData{ ud_put_nonbinding_name = put_name } -> put_name bh name
+
+   get bh =
+      case getUserData bh of
+        UserData { ud_get_name = get_name } -> get_name bh
+```
+
+Later, in `GHC.Iface.Binary`, `getWithUserData` and `putWithUserData` structure the header in the way previously described, and initialise the `UserData` functions to write to/read from mutable tables:
```
[the remark this answers is not in the document]
Comment (JoshMeredith) [reply blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md]: Changes incorporated

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md (hunk 3 above)
Comment (JoshMeredith) [reply blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md:173]: updated the advantages/disadvantages section

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoshMeredith) [reply blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md]: Elaborated on certain comparison points to make these claims more explicit

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md (hunk 3 above)
Comment (doyougnu) [reply blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md:173]: Ah the trouble is architectural, because each section doesn't have a summary paragraph you reach a point where you have to describe the adv/disadv of both at the same time. This makes the piece hard to read because as the reader I have to remember about `GHC.Binary` _after_ just reading through the `Objectable` stuff _in order_ to make sense of the comparison. The reason you're having trouble writing this is because once you put the comparisons down on paper you noticed the redundancy and correctly identified it as a writing smell, which it is. 

So the fix is to add the summary paragraphs to each section. Then this section should relate the adv/disadv which have already been stated to larger architectural points. So it should step back and ask "well what do we really need here? What are the engineering constraints and what are our options? Is performance more important than readability and usability here?" The answers to these questions would then inform the decision on which approach to take and why. This all should be done in an intro paragraph to the `Comparison` section:

> So we have seen two approaches for achieving the same behavior and we have a classic engineering decision on our hands. On the one hand we have the GHC approach which provides performance at the cost of type safety, maintainence and readability. In contrast, the GHCJS approach <another-summary...>. Stepping back we should ask what exactly are our constraints. Will this code be used by unknown consumers? Is the maintenance burden a cost worth bearing...

something like that.

File: blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    deps1        = putDepsSection deps0
+    (sti, idx)   = putIndex st0 os
+    symbs        = putSymbolTable sti
+```
+
+## Comparison
+
+### Advantages of the GHC Binary Approach
+
+* Interface file generation appears to be very fast: speculatively looking at the generated core, this could be because the optimiser has a much better time with the style of IO code that is used - rather than a limitation of more abstacted approaches
```
[the remark this answers is not in the document]
Comment (doyougnu) [reply blog/2022-05-24-april-GHCJS-Objectable-vs-GHC-Binary.md]: I think that instead of:

> Lazy ByteStrings are potentially slower than raw binary buffers, however more benchmarking would be required to confirm this

You could just say:

> Lazy ByteStrings are theoretically slower than raw binary buffers, although we did not empirically measure this.
